### PR TITLE
interview pages

### PR DESCRIPTION
**Pull request by elimelt**
Author: Elijah

## What changes were made?

Implemented interview pages

### CODE REVIEW

File: src/components/InterviewPreview.tsx [hunk 1]
```diff
@@ -9,9 +20,26 @@ interface InterviewPreviewProps {
 export const InterviewPreview: React.FC<InterviewPreviewProps> = ({
   interview,
 }) => {
+  const [interviewer, setInterviewer] = useState<Member | null>(null);
```
Comment (ShawnCollinge) [inline src/components/InterviewPreview.tsx:23]: why not use undefined?

File: src/components/InterviewPreview.tsx [hunk 2]
```diff
@@ -23,19 +51,44 @@ export const InterviewPreview: React.FC<InterviewPreviewProps> = ({
...
+                  {`${interviewer?.firstName} ${interviewer?.lastName}` ||
+                    'Loading...'}
+                </Text>
+                <Badge colorScheme="blue">Interviewer</Badge>
+              </Box>
+            </Flex>
+            <Flex alignItems="center">
+              <Box textAlign="right" mr={3}>
+                <Text fontWeight="bold" color={textColor}>
+                  {`${interviewee?.firstName} ${interviewee?.lastName}` ||
```
Comment (ShawnCollinge) [inline src/components/InterviewPreview.tsx:73]: I don't think you need "?" here as this only loads when it's true.

File: src/components/InterviewPreview.tsx [hunk 3]
```diff
@@ -23,19 +51,44 @@ export const InterviewPreview: React.FC<InterviewPreviewProps> = ({
...
-      </Flex>
-      <Text mt={2}>Interviewer: {interview.interviewer.username}</Text>
-      <Text>
-        Date: {new Date(interview.dateEffective).toLocaleDateString()}
-      </Text>
+      {interviewer && interviewee ? (
+        <>
+          <Flex justifyContent="space-between" alignItems="center">
+            <Flex alignItems="center">
+              <Avatar size="md" name={interviewer?.firstName} mr={3} />
```
Comment (ShawnCollinge) [inline src/components/InterviewPreview.tsx:61]: Same as here.

File: src/pages/ViewInterviewPage.tsx [hunk 4]
```diff
@@ -1,18 +1,26 @@
...
   getBehavioralQuestionsForInterview,
 } from '../services/question';
 import { getInterviewById } from '../services/interview';
 import { useAuth } from '../hooks/useAuth';
+import { getMemberById } from '../services/directory';
 
 export const ViewInterviewPage: React.FC = () => {
   const { interviewId } = useParams<{ interviewId: string }>();
   const [interview, setInterview] = useState<Interview | null>(null);
+  const [interviewer, setInterviewer] = useState<Member | null>(null);
```
Comment (ShawnCollinge) [inline src/pages/ViewInterviewPage.tsx:22]: Again, why not use undefined instead of null

File: src/services/interview.ts [hunk 5]
```diff
@@ -1,39 +1,58 @@
 import { DetailedResponse, Interview, InterviewAvailability } from '../types';
 import api from './api';
 
+function deserializeInterview(data: any): Interview {
```
Comment (ShawnCollinge) [inline src/services/interview.ts:4]: You might want to use unknown or create a type instead of any in case there are issues with the data. Also why is this necessary, looks like you're just renaming fields.

File: src/services/question.ts [hunk 6]
```diff
@@ -4,6 +4,7 @@ import { TechnicalQuestion, BehavioralQuestion } from '../types';
 export async function getTechnicalQuestionsForInterview(
   _interviewId: string
 ): Promise<TechnicalQuestion[]> {
+  return [];
```
Comment (ShawnCollinge) [inline src/services/question.ts:7]: Maybe put a todo here.

File: src/services/interview.ts [hunk 7]
```diff
@@ -1,39 +1,58 @@
...
+    dateCompleted: data.date_completed,
+    interviewer: data.interviewer,
+    interviewee: data.interviewee,
+    status: data.status,
+    technicalQuestion: data.technical_question,
+    behavioralQuestions: data.behavioral_questions,
+  };
+}
+
+function deserializeInterviewAvailability(data: any): InterviewAvailability {
```
Comment (ShawnCollinge) [inline src/services/interview.ts:17]: You should create types instead of using any.

File: src/services/directory.ts [hunk 8]
```diff
@@ -2,6 +2,29 @@ import { Member } from '../types';
 import api from './api';
 import { devPrint } from '../components/utils/RandomUtils';
 
+function deserializeMember(data: any): Member {
```
Comment (ShawnCollinge) [inline src/services/directory.ts:5]: I'm not sure what's going on here, you're just renaming user to id, and first_name to firstName. Most of the fields are the same. Why are we doing this?

File: src/components/InterviewPreview.tsx [hunk 9]
```diff
@@ -23,19 +51,44 @@ export const InterviewPreview: React.FC<InterviewPreviewProps> = ({
...
+            </Flex>
+            <Flex alignItems="center">
+              <Box textAlign="right" mr={3}>
+                <Text fontWeight="bold" color={textColor}>
+                  {`${interviewee?.firstName} ${interviewee?.lastName}` ||
+                    'Loading...'}
+                </Text>
+                <Badge colorScheme="green">Interviewee</Badge>
+              </Box>
+              <Avatar size="md" name={interviewee?.firstName} />
```
Comment (ShawnCollinge) [inline src/components/InterviewPreview.tsx:78]: Same as here

File: src/services/interview.ts (hunk 5 above)
Comment (elimelt) [reply src/services/interview.ts:4]: I will add a type. I essentially am just renaming types, but i didn't want to rewrite the code to do so over and over for each api call

File: src/services/directory.ts (hunk 8 above)
Comment (elimelt) [reply src/services/directory.ts:5]: Our domain objects don't necessarily need to map 1:1 from frontend to backend, and likely won't as we add functionality to either end. In this case the only difference is camel vs snake case, but if we make any additional changes to api layer or frontend, it's nice to only need to update it in one place (the deserializer) as opposed to everywhere we make an api call. Any alternative patterns you suggest using to accomplish this?

Comment (ShawnCollinge) [reply src/services/directory.ts]: to be consistent why not just return the data how you want it with the api call? Create a function called `getMember()` or something that does the api call and returns the data how you want it. 

It's probably better to create functions that make the calls anyways. `getMember()` returns a user object or throws an error.

Comment (elimelt) [reply src/services/directory.ts]: Is this not exactly what the functions in the file are doing? The problem is that multiple functions will inevitably return the same type of object, and so if and when we change what's returned from the backend we will need to then update every function that returns a member as opposed to just updating the serializer.

File: src/services/directory.ts [hunk 10]
```diff
@@ -1,8 +1,9 @@
-import { Member } from '../types';
+import { Member, RawMemberData } from '../types';
 import api from './api';
 import { devPrint } from '../components/utils/RandomUtils';
+import { AxiosResponse } from 'axios';
 
-function deserializeMember(data: any): Member {
+function deserializeMember(data: RawMemberData): Member {
```
Comment (ShawnCollinge) [inline src/services/directory.ts:6]: You can also simplify this as:
```
function deserializeMember({
  user: id,
  first_name: firstName,
  last_name: lastName,
  grad_date: gradDate,
  discord_username: discordUsername,
  resume_url: resumeUrl,
  discord_id: discordId,
  ...rest
}: RawMemberData): Member {
  return {
    id,
    firstName,
    lastName,
    gradDate,
    discordUsername,
    resumeUrl,
    discordId,
    ...rest,
  };
}
```

I really like this pattern personally.

File: src/services/directory.ts (hunk 8 above)
Comment (ShawnCollinge) [reply src/services/directory.ts]: I see, yeah that's fair. I didn't look that closely right now before making that comment so that's perfect.

**Review by ShawnCollinge — APPROVED** [review]
LGTM if you want to change the pattern as I commented might be better but it's up to you.

File: src/services/directory.ts (hunk 10 above)
Comment (elimelt) [reply src/services/directory.ts:6]: oh yeah that looks nice. Also I'm planning on moving deserializers outside of the services module in an subsequent PR so they can be shared FWIW

File: src/types.ts [hunk 11]
```diff
@@ -80,3 +80,40 @@ export interface BehavioralQuestion {
...
+  technical_question: TechnicalQuestion;
+  behavioral_questions: BehavioralQuestion[];
+}
+
+export interface RawInterviewAvailabilityData {
+  user_id: number;
+  availability: boolean[][];
+}
+
+export interface RawMemberData {
```
Comment (ShawnCollinge) [inline src/types.ts:100]: Are these fields all guaranteed? I think some can be undefined so you might want to put them as field?: string. 

Also, this pattern might be nice:
```
interface Member {
  id: number; // user id
  username: string;
  created: string; // ISO 8601 date string
  email: string;
  role: string;
  firstName: string;
  lastName: string;
  preview?: string;
  major?: string;
  gradDate?: string; // ISO 8601 date string or null
  discordUsername: string;
  linkedin?: SocialField;
  github?: SocialField;
  leetcode?: SocialField;
  resumeUrl?: string;
  local?: string;
  bio?: string;
  discordId: number;
}

type RawMemberData = Omit<BaseMember, 'firstName' | 'lastName' | 'gradDate' | 'discordUsername' | 'resumeUrl' | 'discordId'> & {
  user: number;
  first_name: string;
  last_name: string;
  grad_date?: string;
  discord_username: string;
  resume_url?: string;
  discord_id: number;
};
```

also, you can set the dates to `DateTime` types if that makes things easier instead of string.

Comment (elimelt) [reply src/types.ts:100]: good call on optionals. I think for now I'm going to keep it simple and just write out the explicit mapping between names, but we can move to using Omit if it ends up growing too large

Comment (elimelt) [reply src/types.ts:100]: Also will add date parsing to deserializers when I consolidate them outside of services module

**Comment by elimelt** [comment]
merging but I won't rereq

**Review by ShawnCollinge — APPROVED** [review]
